cdc/sink/codec/canal_flat.go

```diff
@@ -0,0 +1,238 @@
+	}
+}
+
+type canalFlatMessage struct {
+	ID            int64                    `json:"id"`
+	Schema        string                   `json:"database"`
+	Table         string                   `json:"table"`
+	PKNames       []string                 `json:"pkNames"`
+	IsDDL         bool                     `json:"isDdl"`
+	EventType     string                   `json:"type"`
+	ExecutionTime int64                    `json:"es"`
+	BuildTime     uint64                   `json:"ts"`
+	Query         string                   `json:"sql"`
+	SQLType       map[string]int32         `json:"sqlType"`
+	MySQLType     map[string]string        `json:"mysqlType"`
+	Data          []map[string]interface{} `json:"data"`
+	Old           []map[string]interface{} `json:"old"`
+	tikvTs        uint64
+}
+
+func (c *CanalFlatEventBatchEncoder) newFlatMessageForDML(e *model.RowChangedEvent) (*canalFlatMessage, error) {
+	eventType := convertRowEventType(e)
+	header := c.builder.buildHeader(e.CommitTs, e.Table.Schema, e.Table.Table, eventType, 1)
+	rowData, err := c.builder.buildRowData(e)
+	if err != nil {
+		return nil, cerrors.WrapError(cerrors.ErrCanalEncodeFailed, err)
+	}
+
+	pkCols := e.PrimaryKeyColumns()
+	pkNames := make([]string, len(pkCols))
+	for i := range pkNames {
+		pkNames[i] = pkCols[i].Name
+	}
+
+	var nonTrivialRow []*canal.Column
+	if e.IsDelete() {
+		nonTrivialRow = rowData.BeforeColumns
+	} else {
+		nonTrivialRow = rowData.AfterColumns
+	}
+
+	sqlType := make(map[string]int32, len(nonTrivialRow))
+	mysqlType := make(map[string]string, len(nonTrivialRow))
+	for i := range nonTrivialRow {
+		sqlType[nonTrivialRow[i].Name] = nonTrivialRow[i].SqlType
+		mysqlType[nonTrivialRow[i].Name] = nonTrivialRow[i].MysqlType
+	}
+
+	var (
+		data    map[string]interface{}
+		oldData map[string]interface{}
+	)
+
+	if len(rowData.BeforeColumns) > 0 {
+		oldData = make(map[string]interface{}, len(rowData.BeforeColumns))
+		for i := range rowData.BeforeColumns {
+			if !rowData.BeforeColumns[i].GetIsNull() {
+				oldData[rowData.BeforeColumns[i].Name] = rowData.BeforeColumns[i].Value
+			} else {
+				oldData[rowData.BeforeColumns[i].Name] = nil
+			}
+		}
+	}
+
+	if len(rowData.AfterColumns) > 0 {
+		data = make(map[string]interface{}, len(rowData.AfterColumns))
+		for i := range rowData.AfterColumns {
+			if !rowData.AfterColumns[i].GetIsNull() {
+				data[rowData.AfterColumns[i].Name] = rowData.AfterColumns[i].Value
+			} else {
+				data[rowData.AfterColumns[i].Name] = nil
+			}
+		}
+	} else {
+		// The event type is DELETE
+		// The following line is important because Alibaba's adapter expects this, and so does Flink.
+		data = oldData
+	}
+
+	ret := &canalFlatMessage{
+		ID:            0,
+		Schema:        header.SchemaName,
+		Table:         header.TableName,
+		PKNames:       pkNames,
+		IsDDL:         false,
+		EventType:     header.GetEventType().String(),
+		ExecutionTime: header.ExecuteTime,
+		BuildTime:     0,
```

Comment:
what does this field mean and why it is zero

Reply:
It means the time when the Canal message is built by TiCDC (or binlog parser or any similar facility). It is not used in the Canal Protobuf implementation, and Alibaba's consumer does not care about it.

Reply:
and why `ID` is zero? Is the `ID` cared by consumer?